Add instructions for Java app servers (#3459)
open-telemetry/opentelemetry.io: content/en/docs/instrumentation/java/automatic/server-config.md

---
title: Application server configuration

> linkTitle: App server config

Reviewer: Can we shorten this a bit more (given that this page is quite nested in the left nav)?

```suggestion
linkTitle: Server config
```

description: Learn how to define agent paths for Java application servers
weight: 15
cSpell:ignore: asadmin binsetenv binstart Glassfish Payara setenv
---

When instrumenting an application running on an application server, you've to
define the path to the OpenTelemetry Java agent. The way to define the path
differs from server to server.

## JBoss EAP / WildFly

To define the path to the Java agent, add the `javaagent` argument at the end of
the standalone configuration file:

> {{< tabpane text=true >}}

Reviewer: ```suggestion
{{< tabpane text=true persist=lang >}}

```

```shell
java -javaagent:/path/to/opentelemetry-javaagent.jar -jar start.jar
```

If you use the `jetty.sh` file to start Jetty, add the following line to the
\<jetty_home\>/bin/jetty.sh file:

```shell
JAVA_OPTIONS="${JAVA_OPTIONS} -javaagent:/path/to/opentelemetry-javaagent.jar"
```

If you use the start.ini file to define JVM arguments, add the `javaagent`
argument after the `--exec` option:

```ini
#===========================================================
# Sample Jetty start.ini file
#-----------------------------------------------------------
--exec
-javaagent:/path/to/opentelemetry-javaagent.jar
```

## Glassfish / Payara

Add the path to the Java agent using the `asadmin` tool:

{{< tabpane text=true >}} {{% tab Linux %}}

```sh
<server_install_dir>/bin/asadmin create-jvm-options "-javaagent\:/path/to/opentelemetry-javaagent.jar"
```

{{% /tab %}} {{% tab Windows %}}

```powershell
<server_install_dir>\bin\asadmin.bat create-jvm-options '-javaagent\:<Drive>\:\\path\\to\\opentelemetry-javaagent.jar'
```

{{% /tab %}} {{< /tabpane >}}

You can also add the `-javaagent` argument from the Admin Console. For example:

> 1.  Open the GlassFish Admin Console at `http://localhost:4848`.
> 
> 2.  Go to **Configurations > server-config > JVM Settings**.
> 
> 3.  Select **JVM Options > Add JVM Option**.
> 
> 4.  Enter the path to the agent:
> 
>     `-javaagent:/path/to/opentelemetry-javaagent.jar`
> 
> 5.  **Save** and restart the server.

Reviewer: Ensure that the rendered list is compact by removing blank lines between entries:

```suggestion
1.  Open the GlassFish Admin Console at <http://localhost:4848>.
2.  Go to **Configurations > server-config > JVM Settings**.
3.  Select **JVM Options > Add JVM Option**.
4.  Enter the path to the agent:
    `-javaagent:/path/to/opentelemetry-javaagent.jar`
5.  **Save** and restart the server.
```

Make sure that the domain.xml file in your domain directory contains a
`<jmv-options>` entry for the agent.

## Tomcat / TomEE

Add the path to the Java agent to your startup script:

```shell
# Linux: Add the following line to \<tomcat_home\>/bin/setenv.sh
CATALINA_OPTS="$CATALINA_OPTS -javaagent:/path/to/opentelemetry-javaagent.jar"

# Windows: Add the following line to \<tomcat_home\>binsetenv.bat
set CATALINA_OPTS=%CATALINA_OPTS% -javaagent:"<Drive>:\path\to\opentelemetry-javaagent.jar"

> ```

Reviewer: You might be working on this already, but I thought that I'd mention it just in case: update to use tabs here.

## WebLogic

Add the path to the Java agent to your domain startup script:

```shell
# Linux: Add the following line to \<domain_home\>/bin/startWebLogic.sh
export JAVA_OPTIONS="$JAVA_OPTIONS -javaagent:/path/to/opentelemetry-javaagent.jar"

# Linux: Add the following line to \<domain_home\>binstartWebLogic.cmd
set JAVA_OPTIONS=%JAVA_OPTIONS% -javaagent:"<Drive>:\path\to\opentelemetry-javaagent.jar"

> ```

Reviewer: Update to use tabs here.

For managed server instances, add the `-javaagent` argument using the admin
console.

## WebSphere Liberty Profile

Add the path to the Java agent to the `jvm.options` file:

```shell
# Single server: Edit ${server.config.dir}/jvm.options
# All servers: Edit ${wlp.install.dir}/etc/jvm.options
-javaagent:/path/to/opentelemetry-javaagent.jar

> ```

Reviewer: ~~~suggestion
Add the path to the Java agent to the `jvm.options` file. For a single server,
edit `${server.config.dir}/jvm.options`, and for all servers, edit
`${wlp.install.dir}/etc/jvm.options`:

```ini
-javaagent:/path/to/opentelemetry-javaagent.jar
```
~~~

Restart the server after saving the file.

## WebSphere Traditional

Open the WebSphere Admin Console and follow these steps:

> 1.  Navigate to **Servers > Server type > WebSphere application servers**.
> 
> 2.  Select the server.
> 
> 3.  Go to **Java and Process Management > Process Definition**.
> 
> 4.  Select **Java Virtual Machine**.
> 
> 5.  In **Generic JVM arguments**, enter the path to the agent:
> 
>     ```bash
>     -javaagent:/path/to/opentelemetry-javaagent.jar
>     ```
> 
> 6.  Save the configuration and restart the server.

Reviewer: Ensure that the rendered list is compact by removing blank lines between entries:

```suggestion
1.  Navigate to **Servers > Server type > WebSphere application servers**.
2.  Select the server.
3.  Go to **Java and Process Management > Process Definition**.
4.  Select **Java Virtual Machine**.
5.  In **Generic JVM arguments**, enter the path to the agent:
    ```bash
    -javaagent:/path/to/opentelemetry-javaagent.jar
    ```
6.  Save the configuration and restart the server.
```

---
title: Application server configuration

> linkTitle: Server config

Reviewer 2: ```suggestion
linkTitle: Application server
```

For me "application server" is something different to just "server", for a Java developer this might be less confusing this way. I would even be OK with `Application server config` (or `App server config`)

Reviewer: Remember that the linkTitle is only used in the left nav. When someone visits the page, they see the full title.

That being said, if you still feel that `Server config` isn't an appropriate title, among the suggestions you make, I'd vote for `App server config`.

Reviewer 2: Then let's go for `App server config` 👍

description: Learn how to define agent paths for Java application servers
weight: 15
cSpell:ignore: asadmin binsetenv binstart Glassfish Payara setenv
---

> When instrumenting an application running on an application server, you've to

Reviewer 2: ```suggestion
When instrumenting an app running on an Java application server, you've to
```

> When instrumenting an application running on an application server, you've to
> define the path to the OpenTelemetry Java agent. The way to define the path
> differs from server to server.

Reviewer 3: ```suggestion
When instrumenting an application running on an application server with a Java Agent, you must
add the `javaagent` path to the OpenTelemetry Java agent to the JVM arguments. The way to do this
differs from server to server.
```

> ## JBoss EAP / WildFly

Reviewer 3: I think it would be nice to start with the generic option: `JAVA_TOOLS_OPTIONS`
https://docs.oracle.com/javase/8/docs/technotes/guides/troubleshoot/envvars002.html

This should help with many situations that aren't explicitly defined here.

Author: Added!

```

{{% /tab %}} {{% tab Windows %}}

```powershell
<server_install_dir>\bin\asadmin.bat create-jvm-options '-javaagent\:<Drive>\:\\path\\to\\opentelemetry-javaagent.jar'
```

{{% /tab %}} {{< /tabpane >}}

You can also add the `-javaagent` argument from the Admin Console. For example:

1.  Open the GlassFish Admin Console at <http://localhost:4848>.
2.  Go to **Configurations > server-config > JVM Settings**.
3.  Select **JVM Options > Add JVM Option**.
4.  Enter the path to the agent:
    `-javaagent:/path/to/opentelemetry-javaagent.jar`
5.  **Save** and restart the server.

Make sure that the domain.xml file in your domain directory contains a
`<jmv-options>` entry for the agent.

## Tomcat / TomEE

Add the path to the Java agent to your startup script:

{{< tabpane text=true persist=lang >}}

{{% tab header="Linux (`<tomcat_home>/bin/setenv.sh`)" lang=Linux %}}

```sh
CATALINA_OPTS="$CATALINA_OPTS -javaagent:/path/to/opentelemetry-javaagent.jar"
```

{{% /tab %}}
{{% tab header="Windows (`<tomcat_home>\bin\setenv.bat`)" lang=Windows %}}

```bat
set CATALINA_OPTS=%CATALINA_OPTS% -javaagent:"<Drive>:\path\to\opentelemetry-javaagent.jar"
```

> {{% /tab %}} {{< /tabpane >}}

Reviewer: Here's an example of how to put the file path in the content, and 

~~~suggestion
{{< tabpane text=true langEqualsHeader=true >}} {{% tab header=Linux %}}

```sh
# <tomcat_home>/bin/setenv.sh
CATALINA_OPTS="$CATALINA_OPTS -javaagent:/path/to/opentelemetry-javaagent.jar"
```

{{% /tab %}} {{% tab header=Windows %}}

```bat
:: <tomcat_home>\bin\setenv.bat
set CATALINA_OPTS=%CATALINA_OPTS% -javaagent:"<Drive>:\path\to\opentelemetry-javaagent.jar"
```

{{% /tab %}} {{< /tabpane >}}
~~~

Author: I went for re-adding the paths as comment in the code snippet, which I think is the most visible way.

## WebLogic

Add the path to the Java agent to your domain startup script:

{{< tabpane text=true persist=lang >}}

{{% tab header="Linux (`<domain_home>/bin/startWebLogic.sh`)" lang=Linux %}}

```sh
export JAVA_OPTIONS="$JAVA_OPTIONS -javaagent:/path/to/opentelemetry-javaagent.jar"
```

{{% /tab %}}
{{% tab header="Windows (`<domain_home>\bin\startWebLogic.cmd`)" lang=Windows %}}

```bat
set JAVA_OPTIONS=%JAVA_OPTIONS% -javaagent:"<Drive>:\path\to\opentelemetry-javaagent.jar"
```

> {{% /tab %}} {{< /tabpane >}}

Reviewer 4: Unfortunately, this renders poorly in the live site. Is there a way to shorten the text in the tabs?

<img width="766" alt="image" src="https://github.com/open-telemetry/opentelemetry.io/assets/6309070/fbe4cf0d-0e5b-46a9-bbc1-a8b028c13ef3">

Reviewer: Yeah, when the tab name is too wide, this isn't optimal on mobile. I think that we do in other pages is add the path in the tab content.

Here's an example of how to fix this: https://github.com/open-telemetry/opentelemetry.io/pull/3459/files#r1383733272

For managed server instances, add the `-javaagent` argument using the admin
console.

## WebSphere Liberty Profile

Add the path to the Java agent to the `jvm.options` file. For a single server,
edit `${server.config.dir}/jvm.options`, and for all servers, edit
`${wlp.install.dir}/etc/jvm.options`:

```ini
-javaagent:/path/to/opentelemetry-javaagent.jar
```

Restart the server after saving the file.

## WebSphere Traditional

Open the WebSphere Admin Console and follow these steps:

> 
> 1.  Navigate to **Servers > Server type > WebSphere application servers**.

Reviewer: Let's quiet the markdownlinter in this specific case:

```suggestion

<!-- markdownlint-disable blanks-around-fences -->

1.  Navigate to **Servers > Server type > WebSphere application servers**.
```

Reviewer: Of course, the other option is to avoid the code block in the list.

---
title: Application server configuration
linkTitle: Server config
description: Learn how to define agent paths for Java application servers
weight: 15
cSpell:ignore: asadmin binsetenv binstart Glassfish Payara setenv
---

When instrumenting an application running on an application server, you've to
define the path to the OpenTelemetry Java agent. The way to define the path
differs from server to server.

## JBoss EAP / WildFly

To define the path to the Java agent, add the `javaagent` argument at the end of
the standalone configuration file:

{{< tabpane text=true persist=lang >}}

> {{% tab header="Linux (standalone.conf)" lang=Linux %}}

Reviewer: ```suggestion
{{% tab header="Linux (`standalone.conf`)" lang=Linux %}}
```

```sh
JAVA_OPTS="$JAVA_OPTS -javaagent:/path/to/opentelemetry-javaagent.jar"
```

> {{% /tab %}} {{% tab header="Windows (standalone.conf.bat)" lang=Windows %}}

Reviewer: ```suggestion
{{% /tab %}} {{% tab header="Windows (`standalone.conf.bat`)" lang=Windows %}}
```

# Sample Jetty start.ini file
#-----------------------------------------------------------
--exec
-javaagent:/path/to/opentelemetry-javaagent.jar
```

## Glassfish / Payara

Add the path to the Java agent using the `asadmin` tool:

{{< tabpane text=true >}} {{% tab Linux %}}

```sh
<server_install_dir>/bin/asadmin create-jvm-options "-javaagent\:/path/to/opentelemetry-javaagent.jar"
```

{{% /tab %}} {{% tab Windows %}}

```powershell
<server_install_dir>\bin\asadmin.bat create-jvm-options '-javaagent\:<Drive>\:\\path\\to\\opentelemetry-javaagent.jar'
```

{{% /tab %}} {{< /tabpane >}}

You can also add the `-javaagent` argument from the Admin Console. For example:

1.  Open the GlassFish Admin Console at <http://localhost:4848>.
2.  Go to **Configurations > server-config > JVM Settings**.
3.  Select **JVM Options > Add JVM Option**.
4.  Enter the path to the agent:
    `-javaagent:/path/to/opentelemetry-javaagent.jar`
5.  **Save** and restart the server.

Make sure that the domain.xml file in your domain directory contains a
`<jmv-options>` entry for the agent.

## Tomcat / TomEE

Add the path to the Java agent to your startup script:

{{< tabpane text=true persist=lang >}}

> {{% tab header="Linux (\<tomcat_home\>/bin/setenv.sh)" lang=Linux %}}

Reviewer: ```suggestion
{{% tab header="Linux (`<tomcat_home>/bin/setenv.sh`)" lang=Linux %}}
```

```sh
CATALINA_OPTS="$CATALINA_OPTS -javaagent:/path/to/opentelemetry-javaagent.jar"
```

{{% /tab %}}

> {{% tab header="Windows (\<tomcat_home\>binsetenv.bat)" lang=Windows %}}

Reviewer: ```suggestion
{{% tab header="Windows (`<tomcat_home>\bin\setenv.bat`)" lang=Windows %}}
```

```bat
set CATALINA_OPTS=%CATALINA_OPTS% -javaagent:"<Drive>:\path\to\opentelemetry-javaagent.jar"
```

{{% /tab %}} {{< /tabpane >}}

## WebLogic

Add the path to the Java agent to your domain startup script:

{{< tabpane text=true persist=lang >}}

> {{% tab header="Linux (\<domain_home\>/bin/startWebLogic.sh)" lang=Linux %}}

Reviewer: ```suggestion
{{% tab header="Linux (`<domain_home>/bin/startWebLogic.sh`)" lang=Linux %}}
```

```sh
export JAVA_OPTIONS="$JAVA_OPTIONS -javaagent:/path/to/opentelemetry-javaagent.jar"
```

{{% /tab %}}

> {{% tab header="Windows (\<domain_home\>binstartWebLogic.cmd)" lang=Windows %}}

Reviewer: ```suggestion
{{% tab header="Windows (`<domain_home>\bin\startWebLogic.cmd`)" lang=Windows %}}
```

```bat
set JAVA_OPTIONS=%JAVA_OPTIONS% -javaagent:"<Drive>:\path\to\opentelemetry-javaagent.jar"
```

{{% /tab %}} {{< /tabpane >}}

For managed server instances, add the `-javaagent` argument using the admin
console.

## WebSphere Liberty Profile

Add the path to the Java agent to the `jvm.options` file. For a single server,
edit `${server.config.dir}/jvm.options`, and for all servers, edit
`${wlp.install.dir}/etc/jvm.options`:

```ini
-javaagent:/path/to/opentelemetry-javaagent.jar
```

Restart the server after saving the file.

## WebSphere Traditional

Open the WebSphere Admin Console and follow these steps:

1.  Navigate to **Servers > Server type > WebSphere application servers**.
2.  Select the server.
3.  Go to **Java and Process Management > Process Definition**.
4.  Select **Java Virtual Machine**.

> 5.  In **Generic JVM arguments**, enter the path to the agent:
> 
>     ```bash
>     -javaagent:/path/to/opentelemetry-javaagent.jar
>     ```
> 
> 6.  Save the configuration and restart the server.

Reviewer: ```suggestion
5.  In **Generic JVM arguments**, enter the path to the agent:
    ```bash
    -javaagent:/path/to/opentelemetry-javaagent.jar
    ```
6.  Save the configuration and restart the server.
```

Reviewer: LOL, ok. Let me check that out.

---

> title: Application server configuration

Reviewer 3: This currently shows at the top level (in the demo version).  How do we make sure it shows up in `instrumentation/java/automatic/`?  Maybe this is a bug with the demo?

Author: Oh, I think the files moved to /languages, could it be? Checking.

Author: @tylerbenson I've moved the files, check the demo: https://deploy-preview-3459--opentelemetry.netlify.app/docs/languages/java/automatic/server-config/

Reviewer 3: looks much better, thanks!

---
title: Application server configuration
linkTitle: App server config
description: Learn how to define agent paths for Java application servers
weight: 15
cSpell:ignore: asadmin binsetenv binstart Glassfish Payara setenv
---

When instrumenting an app that runs on a Java application server with a Java
agent, you must add the `javaagent` path to the JVM arguments. The way to do
this differs from server to server.

## JBoss EAP / WildFly

You can define the path to the Java agent by adding the `javaagent` argument to
the `JAVA_TOOL_OPTIONS` environment variable. For example:

{{< tabpane text=true persist=lang >}}

{{% tab header="Linux" lang=Linux %}}

```sh
export JAVA_TOOL_OPTIONS="-javaagent:/path/to/opentelemetry-javaagent.jar"
```

{{% /tab %}} {{% tab header="Windows" lang=Windows %}}

```bat
set JAVA_TOOL_OPTIONS="-javaagent:<Drive>:\path\to\opentelemetry-javaagent.jar"
```

> {{% /tab %}} {{< /tabpane >}}

Reviewer 5: See https://github.com/open-telemetry/opentelemetry.io/pull/3459/files#r1396086080
Using `JAVA_TOOL_OPTIONS` can have unintended side effects. I think it is better to use the `JAVA_OPTS` way that follows. If needed using `JAVA_TOOL_OPTIONS` can be in a separate section (that is not related to wildfly)

Author: @laurit Would this do?

```suggestion
You can define the path to the Java agent by adding the `javaagent` argument to
the `JAVA_OPTS` environment variable. For example:

{{< tabpane text=true persist=lang >}}

{{% tab header="Linux" lang=Linux %}}

```sh
export JAVA_OPTS="-javaagent:/path/to/opentelemetry-javaagent.jar"
```

{{% /tab %}} {{% tab header="Windows" lang=Windows %}}

```bat
set JAVA_OPTS="-javaagent:<Drive>:\path\to\opentelemetry-javaagent.jar"
```

{{% /tab %}} {{< /tabpane >}}
```

Reviewer 5: Unfortunately no. The instructions ask user to add this at the end of the `standalone.conf` because in https://github.com/mojavelinux/jboss-as/blob/1eeb03a3fa0e795980a6ed96473df3ad0115902b/build/src/main/resources/bin/standalone.conf#L40 there is a test to check whether `JAVA_OPTS` is set, if not it adds some default arguments. When you set `JAVA_OPTS` outside the script setting the default options will be skipped. Unless you also want to explain that this is going to happen I'd suggest not including this in the documentation.

Author: I've updated the PR to just leave the standalone method. PTAL
